pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -610,30 +674,37 @@ func (gc *GarbageCollector) attemptToDeleteItem(ctx context.Context, item *node)
 			// otherwise, default to background.
 			policy = metav1.DeletePropagationBackground
 		}
-		klog.V(2).InfoS("Deleting object", "object", klog.KRef(item.identity.Namespace, item.identity.Name),
-			"objectUID", item.identity.UID, "kind", item.identity.Kind, "propagationPolicy", policy)
+		logger.V(2).Info("Deleting object",
+			"object", klog.KRef(item.identity.Namespace, item.identity.Name),
+			"objectUID", item.identity.UID,
```

Comment:
Elsewhere, this would have been `"item", item`. Should this here be changed to match that or should all other lines be matched to this here ("object" + "objectUID")?

Have you checked that `"item", item.identity` leads to a useful log entry? In other words, what is the output?

If it's useful, then we should use a single key ("item") with a more complex struct (`item.identity`).

Sorry for all the back and forth on this.

Reply:
I'll check and report back

Reply:
`identity` is an `objectReference`, which has a `String()` implementation:

```go
return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
```

Reply:
That explains why this code here doesn't use `item.identity` as value.

But we can do better than this. A single key/value pair is clearly the right solution, we just need to ensure that it gets logged well in text *and* JSON format. Let me try something...

Reply:
You can add this commit to your branch and then pass a single key/value pair. Note that there are other places which log individual values, for example "GraphBuilder process object" in graph_builder.go.

```patch
From 91bef2448f6f0a88f140a49d60795b30ae7f6a88 Mon Sep 17 00:00:00 2001
From: Patrick Ohly <patrick.ohly@intel.com>
Date: Mon, 6 Mar 2023 21:37:31 +0100
Subject: [PATCH 1/1] garbagecollector: structured logging of objectReference

When using JSON as output format, we want objectReference values to be
represented as a struct. For example, "item" is such a value:

{"ts":1678135015708.349,"caller":"garbagecollector/garbagecollector.go:595","msg":"classify object references","v":5,"item":{"name":"dra-test-driver-g4tkd","namespace":"dra-1830","apiVersion":"v1","uid":"c3f88616-7282-488c-887c-3f04291e6f4f"},"solid":null,"dangling":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"dra-test-driver","uid":"dbe9a90c-9dfd-4ad0-8395-e5fa228f9851","controller":true,"blockOwnerDeletion":true}],"waitingForDependentsDeletion":null}
---
 pkg/controller/garbagecollector/graph.go | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/pkg/controller/garbagecollector/graph.go b/pkg/controller/garbagecollector/graph.go
index b2fd8002d5c..de017b4c40b 100644
--- a/pkg/controller/garbagecollector/graph.go
+++ b/pkg/controller/garbagecollector/graph.go
@@ -20,6 +20,8 @@ import (
 	"fmt"
 	"sync"
 
+	"github.com/go-logr/logr"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 )
@@ -30,10 +32,29 @@ type objectReference struct {
 	Namespace string
 }
 
+// String is used when logging an objectReference in text format.
 func (s objectReference) String() string {
 	return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
 }
 
+// MarshalLog is used when logging an objectReference in JSON format.
+func (s objectReference) MarshalLog() interface{} {
+	return struct {
+		Name       string    `json:"name"`
+		Namespace  string    `json:"namespace"`
+		APIVersion string    `json:"apiVersion"`
+		UID        types.UID `json:"uid"`
+	}{
+		Namespace:  s.Namespace,
+		Name:       s.Name,
+		APIVersion: s.APIVersion,
+		UID:        s.UID,
+	}
+}
+
+var _ fmt.Stringer = objectReference{}
+var _ logr.Marshaler = objectReference{}
+
 // The single-threaded GraphBuilder.processGraphChanges() is the sole writer of the
 // nodes. The multi-threaded GarbageCollector.attemptToDeleteItem() reads the nodes.
 // WARNING: node has different locks on different fields. setters and getters
-- 
2.30.2

```

Reply:
Thanks for this! Just pushed with this and more logging changes to hopefully do what you're looking for 😄 . Note, I didn't touch `GraphBuilder process object` because that is the data that comes from an event, before there is a node. I think it makes sense to keep these separate.